Re-enabling type safety for the SPP

Msarawan opened the pull request:
Enabled strict TypeScript checking across the entire codebase by resolving all remaining type errors in `UI/form` components and consolidating per-reducer type checks into a single typecheck.

fixes: [issue#170](https://github.com/arkmq-org/activemq-artemis-self-provisioning-plugin/issues/170)

lavocatt reviewed (changes requested):
Thanks for all the work on this — I know strict‑mode is a big chunk of work with lots of value and I want it to land. But the current diff is too large and it introduces behavior changes/regressions we can’t safely verify with current tests.

I’d like to keep your work intact and merge it in small, safe pieces.

## Why I’m nervous about the current diff

I found regressions:

- Clearing secrets no longer works (reducers return early on secret: undefined).
- cert‑manager issuer updates can wipe existing resourceTemplates/annotations.
- Some validation helpers now return non‑booleans (areMandatoryValuesSet713, areMandatoryValuesSetRestricted), which can silently pass checks.

And I'm confident there are more, and I'm afraid to miss them.

# The plan:

Your work is valuable — I just need it to land safely and predictably. In smaller pieces.

## List of PRs

1) Reducers 7.12
   - `src/reducers/7.12/reducer.ts`
   - `src/reducers/7.12/reducer.test.ts`
   - `src/reducers/7.12/import-types.ts`

2) Reducers 7.13
   - `src/reducers/7.13/reducer.ts`
   - `src/reducers/7.13/reducer.test.ts`

3) Restricted reducer
   - `src/reducers/restricted/reducer.ts`
   - `src/reducers/restricted/reducer.test.ts`

4) Global reducer
   - `src/reducers/reducer.ts`
   - `src/reducers/reducer.test.ts`

5) UI/Forms
   - Shared components (e.g., `FormView`, `CertSecretSelector`, etc.)

---

## Checklist per PR

- Scope: Only files in this slice (no cross‑cutting edits)
- Strict fix only: No behavior change unless explicitly called out
- Tests: Update/add tests only for this slice
- Regression scan:  
  - clear/empty values still work  
  - no “early return” removes side effects  
  - no data structure resets (arrays/objects) that wipe existing config  
- Typecheck: run the strict config for this slice  
  - `yarn tsc -p tsconfig.strict.reducers.json`

---

## Review expectations
- Include a short “Behavior changes? Yes/No” note in each PR.
- If Yes, list the exact behavior change and why it’s safe.

---

## How to enable strict only for a subpart (example: reducers)

Add tsconfig.strict.reducers.json:

```
  {
    "extends": "./tsconfig.json",
    "compilerOptions": {
      "strict": true,
      "noEmit": true
    },
    "include": [
      "src/reducers/**/*.ts",
      "src/reducers/**/*.tsx"
    ]
  }
```

Then run: `yarn tsc -p tsconfig.strict.reducers.json`

We can repeat this per slice and add each to CI.

Msarawan committed:
[#170] Re-enable type safety for the SPP

Enabled strict TypeScript checking across the entire codebase
by resolving all remaining type errors in UI/form components and
consolidating per-reducer type checks into a single typecheck.

lavocatt commented on src/brokers/add-broker/AddBroker.component.tsx:
```diff
@@ -220,7 +220,7 @@ export const AddBroker: FC<AddBrokerPropTypes> = ({
                     setUserWantsToReload(true);
                   } else {
-                    reloadExisting();
+                    reloadExisting?.();
```
Can you explain me this syntax please?

Msarawan replied:
This is optional chaining on a function call. It only calls the function if `reloadExisting` is not undefined or null. It's equivalent to:

if (reloadExisting) {
  reloadExisting();
}

lavocatt commented on src/brokers/add-broker/AddBroker.container.tsx:
```diff
@@ -81,6 +103,9 @@ export const AddBrokerPage: FC = () => {
   }
 
+  const { cr } = brokerModel;
+  if (!cr) return null;
```
Returning null here, this would lead to a strange UI, granted it's not probably going to happen, but what could you do to improve the situation?

Msarawan replied:
we can return a `GenericError` component instead of null. in that way we can keep consistency with all the other places in the codebase.

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/components/CardBrokerPendingMessagesMetrics/CardBrokerPendingMessagesMetrics.container.tsx:
```diff
@@ -15,12 +15,18 @@ type CardBrokerPendingMessagesMetricsContainerProps = {
+  const endTime = Date.now();
+  return [endTime - span, endTime];
+};
+
```
this is defined in multiple places, worth factorizing somewhere.

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/components/ChartCPUUsage/ChartCPUUsage.tsx:
```diff
@@ -82,19 +82,19 @@ export const ChartCPUUsage: FC<ChartCPUUsageProps> = ({
         legendData.push({ name });
         tooltipSeriesNames.push(name);
-      } else {
+      } else if (metric) {
```
what if there's no metrics? Does that result in an error somewhere?

Msarawan replied:
No, it won't cause an error. If metric is missing, the chart line still renders — just without a tooltip label. This is unlikely in practice since Prometheus always returns metrics with labels. The else if (metric) guard is intentional to avoid pushing undefined into the label

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/components/MetricsPolling/MetricsPolling.tsx:
```diff
@@ -60,7 +60,7 @@ export const MetricsPolling: React.FC<MetricsPollingProps> = ({
   });
 
   useEffect(() => {
-    onResult(index, result, loaded, loadError);
+    onResult(index, result ?? null, loaded, loadError);
   }, [index, result, loaded, loadError, onResult]);
 
```
I'm not sure about this, what happens if there's no result? Are we surfacing an error somewhere?

Follow up action is to remove that useEffect, but out of scope for the current PR.

Msarawan replied:
?? null is just a type fix — `usePrometheusPoll`  can return undefined, but onResult expects null. Behavior is the same as main. Errors are already handled via loadError. and missing data shows an empty state via `EmptyStateNoMetricsData` component .

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/components/QueryBrowser/QueryBrowser.tsx:
```diff
@@ -119,8 +119,12 @@ export const QueryBrowser: FC<QueryBrowserProps> = ({
+  const isGraphPoint = (
+    point: GraphDataPoint | undefined,
+  ): point is GraphDataPoint => point !== undefined;
+
```
There are several occurrences of this function definition, worth extracting somewhere in a common place.

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/hooks/useChartWidth.tsx:
```diff
@@ -1,10 +1,10 @@
 
 export const useChartWidth = () => {
-  const ref = useRef<HTMLDivElement>(null);
+  const ref = useRef<HTMLDivElement | null>(null);
```
Why not instead initializing it with an empty ref instead of using the type null?

Msarawan replied:
The | null is needed here because we assign to ref.current directly (line 12). Without it, TypeScript infers a read-only RefObject instead of MutableRefObject, which makes ref.current non-assignable.

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/hooks/useChartWidth.tsx:
```diff
@@ -1,10 +1,10 @@
   const [width, setWidth] = useState<number>();
 
-  const setRef = useCallback((e: HTMLDivElement) => {
+  const setRef = useCallback((e: HTMLDivElement | null) => {
```
same question

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/utils/prometheus.ts:
```diff
@@ -46,11 +46,11 @@ export const parsePrometheusDuration = (duration: string): number => {
-      0,
-    );
+
+    return parts.reduce((sum, p) => {
```
can we keep the type RegExpMatchArray | null with the definition of p please?

lavocatt commented on src/brokers/broker-details/components/Overview/Metrics/utils/units.ts:
```diff
@@ -87,8 +87,8 @@ const convertBaseValueToUnits = (
-  initialUnit: string,
-  preferredUnit: string,
+  initialUnit?: string,
+  preferredUnit?: string,
```
why making these two fields optional?

Msarawan replied:
yeah, it's not needed. i will update the code.

lavocatt commented on src/brokers/broker-details/components/Overview/ConnectivityTester.tsx:
```diff
@@ -386,6 +397,9 @@ export const ConnectivityTester: FC<ConnectivityTesterProps> = ({ cr }) => {
     : '';
 
   const handleCreateClientCert = async () => {
+    if (!namespace) {
+      return;
+    }
```
shouldn't that be an error instead of an empty return?

	
```
      setClientCertStatus({
        status: 'error',
        message: err?.message || t('Failed to create messaging-client-cert.'),
      });
```

lavocatt commented on src/brokers/broker-details/components/Overview/ConnectivityTester.tsx:
```diff
@@ -526,6 +542,9 @@ export const ConnectivityTester: FC<ConnectivityTesterProps> = ({ cr }) => {
     command: string,
     setStatus: (s: ActionState) => void,
   ) => {
+    if (!namespace) {
+      return;
+    }
```
same observation here

lavocatt commented on src/brokers/broker-details/components/Overview/ConnectivityTester.tsx:
```diff
@@ -583,6 +603,9 @@ export const ConnectivityTester: FC<ConnectivityTesterProps> = ({ cr }) => {
   const handleCleanup = async () => {
+    if (!namespace) {
+      return;
+    }
```
same observation here.

lavocatt commented on src/brokers/broker-details/components/Resources/components/ResourcesList.tsx:
```diff
@@ -24,18 +24,17 @@ export const ResourcesTable: FC<ResourcesTableProps> = ({ data }) => {
-  const [activeSortIndex, setActiveSortIndex] = useState<number | null>(null);
+  const [activeSortIndex, setActiveSortIndex] = useState<number | undefined>(
+    undefined,
+  );
```
what about:  `const [activeSortIndex, setActiveSortIndex] = useState<number>(0);` ?

lavocatt commented on src/brokers/broker-details/components/Resources/components/ResourcesList.tsx:
```diff
@@ -24,18 +24,17 @@ export const ResourcesTable: FC<ResourcesTableProps> = ({ data }) => {
-    'asc' | 'desc' | null
-  >(null);
+    'asc' | 'desc' | undefined
+  >(undefined);
```
what about initializing with asc instead?

lavocatt commented on src/brokers/broker-details/components/yaml/Yaml.container.tsx:
```diff
@@ -12,7 +12,23 @@ const YamlContainer: FC = () => {
+        variant={AlertVariant.danger}
+        isInline
+        title={t('Missing required parameters')}
+      >
+        {t('namespace and broker name are required.')}
+      </Alert>
+    );
+  }
+
```
The same component is declared numerous times, worth extracted in a common component

lavocatt commented on src/brokers/update-broker/UpdateBroker.container.tsx:
```diff
@@ -16,46 +16,81 @@ import {
+      </Alert>
+    );
+  }
+
```
Please do not introduce `useEffect` in the code base.
As a reminder: https://react.dev/learn/you-might-not-need-an-effect

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/ListPresets/ListPresets.tsx:
```diff
@@ -26,10 +26,18 @@ const CertManagerPreset: FC<ResourceTemplateProps> = ({ resourceTemplate }) => {
+
+  if (!acceptor || !acceptor.name) {
+    return null;
+  }
```
shall we return null or an error component instead?
remark valid for this return and the one above

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/ListPresets/ListPresets.tsx:
```diff
@@ -86,11 +94,15 @@ const CertManagerPreset: FC<ResourceTemplateProps> = ({ resourceTemplate }) => {
   const { cr } = useContext(BrokerCreationFormState);
+  if (!cr) return null;
+  if (!acceptor) {
+    return <></>;
```
The question we should be asking ourselves, is what does that mean here to get en empty acceptor, shouldn't that be handled on the caller side? No acceptor, then no need to list its presets isn't it?

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/OtherParameters/OtherParameters.tsx:
```diff
@@ -32,8 +32,11 @@ const Param: FC<ParamProps> = ({
   const dispatch = useContext(BrokerCreationFormDispatch);
+  const [newKey, setNewKey] = useState(key);
+  const [newValue, setNewValue] = useState(value);
+  if (!cr) return null;
```
again, return null or an error component?
we could have a pretty generic error component for this situation

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/OtherParameters/OtherParameters.tsx:
```diff
@@ -111,13 +112,14 @@ export const OtherParameters: FC<OtherParametersProps> = ({
   const { t } = useTranslation();
   const { cr } = useContext(BrokerCreationFormState);
+  const dispatch = useContext(BrokerCreationFormDispatch);
+  if (!cr) return null;
```
again, return null or an error component?
we could have a pretty generic error component for this situation

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/PresetAlertPopover/PresetAlertPopover.tsx:
```diff
@@ -22,13 +22,12 @@ export const PresetAlertPopover: FC<PresetCautionProps> = ({
 }) => {
   const { cr } = useContext(BrokerCreationFormState);
   const { t } = useTranslation();
+  if (!cr) return null;
```
again, return null or an error component?
we could have a pretty generic error component for this situation

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/AcceptorConfigPage.tsx:
```diff
@@ -45,6 +45,8 @@ export const AcceptorConfigPage: FC<AcceptorProps> = ({
   const { cr } = useContext(BrokerCreationFormState);
   const dispatch = useContext(BrokerCreationFormDispatch);
+  const { t } = useTranslation();
+  if (!cr) return null;
```
again, return null or an error component?
we could have a pretty generic error component for this situation

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/PresetAlertPopover/PresetAlertPopover.tsx:
```diff
@@ -22,13 +22,12 @@ export const PresetAlertPopover: FC<PresetCautionProps> = ({
+    getCertManagerResourceTemplateFromAcceptor(cr, acceptor) !== undefined;
 
   if (!hasCertManagerPreset) {
     return <></>;
```
we could have a pretty generic error component for this situation

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/AcceptorConfigPage/PresetAlertPopover/PresetAlertPopover.tsx:
```diff
@@ -22,13 +22,12 @@ export const PresetAlertPopover: FC<PresetCautionProps> = ({
-      : false;
+    configType === ConfigType.acceptors &&
+    acceptor !== undefined &&
+    getCertManagerResourceTemplateFromAcceptor(cr, acceptor) !== undefined;
```
are you sure this new condition is semantically equivalent to the one it replaces?

lavocatt replied:
I think it is, but it took me some mental processing.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/ConfigRenamingModal/ConfigRenamingModal.tsx:
```diff
@@ -28,8 +28,12 @@ export const ConfigRenamingModal: FC<ConfigRenamingModalProps> = ({
-  const [validateStatus, setValidateStatus] = useState(null);
+  const [validateStatus, setValidateStatus] = useState<
+    ValidatedOptions | undefined
+  >(undefined);
```
I'm wondering if we shouldn't initialize this with some value compatible with the ValidateOptions type instead of using undefined.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/ConfigRenamingModal/ConfigRenamingModal.tsx:
```diff
@@ -28,8 +28,12 @@ export const ConfigRenamingModal: FC<ConfigRenamingModalProps> = ({
+  >(undefined);
   const { cr } = useContext(BrokerCreationFormState);
+  const [isOpen, setIsOpen] = useState(false);
+  if (!cr) return null;
```
same comment on the need for a generic error component.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/PresetButton/PresetButton.tsx:
```diff
@@ -32,11 +32,11 @@ import { SelectIssuerDrawer } from '../../../../../SelectIssuerDrawer/SelectIssu
 import { ExclamationTriangleIcon } from '@patternfly/react-icons';
 type PreconfigurationButtonProps = {
-  acceptor: Acceptor;
+  acceptor: Acceptor | undefined;
```
I'm not convinced this should accept undefined as a potential type.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/PresetButton/PresetButton.tsx:
```diff
@@ -82,12 +82,13 @@ const AddPresetModal: FC<AddIssuerAnnotationModalProps> = ({
   const [showCertManagerForm, setShowCertManagerForm] = useState(false);
   const { hasCertManager, isLoading: isLoadingCertManagerAvailability } =
     useHasCertManager();
+  if (!cr) return null;
```
same comment on the generic error component being needed.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AcceptorsConfigPage/AcceptorConfigSection/PresetButton/PresetButton.tsx:
```diff
@@ -219,6 +220,11 @@ const AddPresetModal: FC<AddIssuerAnnotationModalProps> = ({
+
+  if (!acceptor) {
+    return null;
+  }
```
By saying that the acceptor is optional, we're creating a bit of a struggle for ourselves here. I think it makes sense to check the presence of the acceptor at the caller level.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AccessControl/SecurityRules.tsx:
```diff
@@ -27,8 +27,11 @@ type SecurityRulesProps = {
   const dispatch = useContext(BrokerCreationFormDispatch);
+  const [newKey, setNewKey] = useState(key);
+  const [newValue, setNewValue] = useState(value);
+  if (!cr) return null;
```
same comment on the generic error component being needed.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AccessControl/SecurityRules.tsx:
```diff
@@ -99,10 +100,10 @@ const SecurityRule: FC<SecurityRulesProps> = ({ name: key, value: value }) => {
   const dispatch = useContext(BrokerCreationFormDispatch);
-
   const [filterValue, setFilterValue] = useState<string>('');
+  if (!cr) return null;
```
same comment on the generic error component being needed.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AccessControl/UserMappings.tsx:
```diff
@@ -127,11 +130,13 @@ export const SelectUserMappings: FC = () => {
+      .filter((name): name is string => Boolean(name))
+      .map((name) => ({
+        value: name,
+        content: name,
```
Map then filter then map feels too verbose to me.

You can use a flatmap instead:

```
return validSecrets.flatMap((jaasConfig) => {
  const name = jaasConfig.metadata?.name;
  
  // Return an array with the object if it's valid, or an empty array to filter it out
  return name ? [{ value: name, content: name }] : [];
});
```

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/AccessControl/UserMappings.tsx:
```diff
@@ -185,12 +190,12 @@ export const SelectUserMappings: FC = () => {
   };
 
   const triggerJaasConfigCreation = () => {
+    const namespace = cr?.metadata?.namespace;
+    if (!namespace) {
+      return;
+    }
```
setAlertSecret() before empty return, because that's an issue if there's no namespace.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/CertSecretSelector/CertSecretSelector.tsx:
```diff
@@ -132,53 +132,60 @@ const GenerateCertificateModal: FC<GenerateCertificateModalProps> = ({
   const dispatch = useContext(BrokerCreationFormDispatch);
   const [selectedIssuer, setSelectedIssuer] = useState<string>('');
   const { cr } = useContext(BrokerCreationFormState);
+  if (!cr) return null;
```
here we could use the generic error component.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/CertSecretSelector/CertSecretSelector.tsx:
```diff
@@ -132,53 +132,60 @@ const GenerateCertificateModal: FC<GenerateCertificateModalProps> = ({
       return;
     }
+    const namespace = cr.metadata?.namespace;
+    if (!namespace) return;
```
here we could use the generic error component.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/CertSecretSelector/CertSecretSelector.tsx:
```diff
@@ -590,6 +600,14 @@ export const CertSecretSelector: FC<CertSecretSelectorProps> = ({
+  const showCertTooltipRef = useRef<HTMLButtonElement>(null);
+  const [parseError, setParseError] = useState(null);
+
+  if (!cr) return null;
```
here we could use the generic error component.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/CertSecretSelector/CertSecretSelector.tsx:
```diff
@@ -665,33 +683,31 @@ export const CertSecretSelector: FC<CertSecretSelectorProps> = ({
-        });
+        const keys = Object.keys(theSecret[0].data);
+        if (keys.length === 0) return;
+        pem = base64.decode(theSecret[0].data[keys[0]]);
```
Here you've changed the control flow. Before we were selecting the last key, now the first key is the one selected.

lavocatt commented on src/shared-components/FormView/BrokerProperties/ConfigurationPage/ConsoleConfigPage/ConsoleConfigPage.tsx:
```diff
@@ -58,6 +58,8 @@ export const ConsoleConfigPage: FC<ConsoleConfigProps> = ({ brokerId }) => {
   };
 
+  const [execOnlyOnce, setExecOnlyOnce] = useState(true);
+  if (!cr) return null;
```
Another reference to the generic error component.

lavocatt commented on src/shared-components/FormView/RestrictedMode/ControlPlane/GenerateRestrictedCertModal.tsx:
```diff
@@ -97,22 +97,28 @@ export const GenerateRestrictedCertModal: FC<
     }
+    const namespace = cr?.metadata?.namespace;
+    if (!namespace) {
+      return;
```
Another reference to the generic error component.

lavocatt commented on src/shared-components/FormView/RestrictedMode/DataPlane/dataPlaneSecretHooks.ts:
```diff
@@ -47,6 +47,9 @@ export const useCreateAmqpsPemSecret = ({
 
   const reconcileAmqps = async (runId: number) => {
+    if (!namespace || !brokerName) {
+      return;
```
Another reference to the generic error component.

lavocatt commented on tsconfig.strict.reducers.json:
```diff
@@ -0,0 +1,8 @@
+    "noEmit": true
+  },
+  "include": ["src/reducers/**/*.ts", "src/reducers/**/*.tsx"]
+}
```
do we need this file?

Msarawan replied:
No , we don't need this file now. i will update the code. thank you!

lavocatt replied:
This file is still here 🙂

lavocatt reviewed (commented):
That's a good piece of work, I've got some comments, if you have questions let me know!

These:
```
  if (!namespace || !brokerName) {
    return null;
  }
```
Should be:
```
  if (!namespace || !brokerName) {
    return GenericErrorComponent;
  }
```

lavocatt reviewed (approved):
LGTM! Congrats for the huge amount of work you've done and for the much improved quality of code.
